src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs

```diff
@@ -408,6 +408,62 @@ internal void ThrowIfCompletedUnsuccessfully()
+        /// <returns>The <see cref="Task"/> representing the asynchronous wait.  It may or may not be the same instance as the current instance.</returns>
+        /// <remarks>
+        /// Timeouts don't <see cref="WaitAsync"/> affect the execution the original task itself; if a <see cref="WaitAsync"/> is canceled or times out on a task that
+        /// hasn't yet completed, the original task may continue to execute even after the <see cref="WaitAsync"/> task has completed.
+        /// Once <see cref="WaitAsync"/> has been called, the original <see cref="ValueTask"/> shouldn't be accessed again, regardless of
+        /// whether the operation completes or times out.
+        /// </remarks>
+        public ValueTask WaitAsync(TimeSpan timeout) =>
+            WaitAsync(timeout, default);
+
+        /// <summary>Gets a <see cref="ValueTask"/> that will complete when this <see cref="ValueTask"/> completes, when the specified timeout expires, or when the specified <see cref="CancellationToken"/> has cancellation requested.</summary>
+        /// <param name="timeout">The timeout after which the <see cref="ValueTask"/> should be faulted with a <see cref="TimeoutException"/> if it hasn't otherwise completed.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
+        /// <returns>The <see cref="Task"/> representing the asynchronous wait.  It may or may not be the same instance as the current instance.</returns>
+        /// <remarks>
+        /// Canceling or timing out the <see cref="WaitAsync"/> doesn't affect the execution of the original task itself; if a <see cref="WaitAsync"/> is canceled or times out on a task that
+        /// hasn't yet completed, the original task may continue to execute even after the <see cref="WaitAsync"/> task has completed.
+        /// Once <see cref="WaitAsync"/> has been called, the original <see cref="ValueTask"/> shouldn't be accessed again, regardless of
+        /// whether the operation completes, times out, or is canceled.
+        /// </remarks>
+        public ValueTask WaitAsync(TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            uint millisecondsTimeout = Task.ValidateTimeout(timeout, ExceptionArgument.timeout);
+
+            if (IsCompleted)
+            {
+                return this;
+            }
+
+            // Perf note: If the ValueTask represents an incomplete operation backed by an IValueTaskSource, this implementation will result in
+            // one more Task allocation than is truly necessary, as it first marshals the IValueTaskSource to a Task via AsTask(), and then uses
+            // the shared WithTimeoutAndCancellation routine on that.  It would be possible to instead have a specialized AsTask(timeout, cancellationToken)
```

Comment:
```suggestion
            // the shared WaitAsync routine on that.  It would be possible to instead have a specialized AsTask(timeout, cancellationToken)
```